docs/src/literate-howto/threaded_assembly.jl

```diff
@@ -1,202 +1,316 @@
+#  - `assembler`: the assembler (which needs to be duplicated because it contains buffers
+#    that are modified during the call to `assemble!`)
+struct ScratchData{CC, CV, T, A}
+    cell_cache::CC
+    cellvalues::CV
     Ke::Matrix{T}
     fe::Vector{T}
-    cellvalues::CV
-    facetvalues::FV
-    global_dofs::Vector{Int}
-    ɛ::Vector{TT}
-    coordinates::Vector{Vec{dim, T}}
-    assembler::AT
-end;
-
-# Each thread need its own CellValues and FacetValues (although, for this example we don't use
-# the FacetValues)
-function create_values(interpolation_space::Interpolation{refshape}, qr_order::Int) where {dim, refshape<:Ferrite.AbstractRefShape{dim}}
-    ## Interpolations and values
-    quadrature_rule = QuadratureRule{refshape}(qr_order)
-    facet_quadrature_rule = FacetQuadratureRule{refshape}(qr_order)
-    cellvalues = [CellValues(quadrature_rule, interpolation_space) for i in 1:Threads.nthreads()];
-    facetvalues = [FacetValues(facet_quadrature_rule, interpolation_space) for i in 1:Threads.nthreads()];
-    return cellvalues, facetvalues
-end;
-
-# Create a `ScratchValues` for each thread with the thread local data
-function create_scratchvalues(K, f, dh::DofHandler{dim}, ip) where {dim}
-    nthreads = Threads.nthreads()
-    assemblers = [start_assemble(K, f) for i in 1:nthreads]
-    cellvalues, facetvalues = create_values(ip, 2)
-
-    n_basefuncs = getnbasefunctions(cellvalues[1])
-    global_dofs = [zeros(Int, ndofs_per_cell(dh)) for i in 1:nthreads]
-
-    fes = [zeros(n_basefuncs) for i in 1:nthreads] # Local force vector
-    Kes = [zeros(n_basefuncs, n_basefuncs) for i in 1:nthreads]
-
-    ɛs = [[zero(SymmetricTensor{2, dim}) for i in 1:n_basefuncs] for i in 1:nthreads]
-
-    coordinates = [[zero(Vec{dim}) for i in 1:length(dh.grid.cells[1].nodes)] for i in 1:nthreads]
-
-    return [ScratchValues(Kes[i], fes[i], cellvalues[i], facetvalues[i], global_dofs[i],
-                         ɛs[i], coordinates[i], assemblers[i]) for i in 1:nthreads]
-end;
-
-# ## Threaded assemble
+    assembler::A
+end
 
-# The assembly function loops over each color and does a threaded assembly for that color
-function doassemble(K::SparseMatrixCSC, colors, grid::Grid, dh::DofHandler, C::SymmetricTensor{4, dim}, ip) where {dim}
+# This constructor will be called within each task to create a independent `ScratchData`
+# object. For `cell_cache`, `Ke`, and `fe` we simply call the constructors to allocate
+# independent objects. For `cellvalues` we use `copy` which Ferrite defines for this
+# purpose. Finally, for the assembler we call `start_assemble` to create a new assembler but
+# note that we set `fillzero = false` because we don't want to risk that a task that starts
+# a bit later will zero out data that another task have already assembled.
+function ScratchData(dh::DofHandler, K::SparseMatrixCSC, f::Vector, cellvalues::CellValues)
+    cell_cache = CellCache(dh)
+    n = ndofs_per_cell(dh)
+    Ke = zeros(n, n)
+    fe = zeros(n)
```

Comment:
Previously we allocated the scratch data in a different way: `fes = [zeros(n_basefuncs) for i in 1:nthreads]` (to avoid cache misses I believe). Is that not needed anymore? This is definitely cleaner.

Reply:
This constructor will be called once per task yes.